Dps sample/documentation (#936)
Azure/azure-sdk-for-c: sdk/samples/iot/provisioning/README.md

The samples' instructions include specifics for both Windows and Linux based systems.  For Linux, the examples are tailored to Debian/Ubuntu environments.  The samples are also designed to work on MacOS systems, but the instructions do not yet include specific command line examples for this environment.

## Key Concepts

Further background on the Azure IoT Client library and key concepts are explained in detail in the [Azure IoT Client README](https://github.com/Azure/azure-sdk-for-c/tree/master/sdk/docs/iot#azure-iot-clients).

## Prerequisites

* Have an [Azure account](https://azure.microsoft.com/en-us/) created.
* Have an [Azure IoT Hub](https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-create-through-portal) created.

> * Have an [Azure IoT Hub Device Provisioning Service (DPS)](https://docs.microsoft.com/en-us/azure/iot-dps/quick-setup-auto-provision) created.

Reviewer: ```suggestion
* Optionally, have an [Azure IoT Hub Device Provisioning Service (DPS)](https://docs.microsoft.com/en-us/azure/iot-dps/quick-setup-auto-provision) created.
```

* For Windows systems, have [Microsoft Visual Studio](https://visualstudio.microsoft.com/downloads/) installed.
* Have [git](https://git-scm.com/download) installed.
* Have [OpenSSL](https://www.openssl.org/source/) installed:

  For Linux based systems (Debian/Ubuntu), we recommend:
  ```bash
  sudo apt-get install libssl-dev
  ```
  For non-Linux based systems, download the [OpenSSL 1.1.1 LTS](https://www.openssl.org/source/openssl-1.1.1g.tar.gz) command line utility and follow the downloaded INSTALL document.
* Have the latest version of [CMake](https://cmake.org/download) installed.  For OS specific installation instructions, follow the downloaded README document.
* Have VCPKG and [Eclipse Paho MQTT C client](https://www.eclipse.org/paho/) installed.  Use the directions [here](https://github.com/Azure/azure-sdk-for-c#development-environment) to download VCPKG and install Paho MQTT.

## Getting Started

### Paho IoT Provisioning (Certificates)

> This [sample](https://github.com/Azure/azure-sdk-for-c/blob/master/sdk/samples/iot/provisioning/src/paho_iot_provisioning_example.c) uses x509 authentication to connect to Azure IoT Hub Device Provisioning Service (DPS).  To easily run this sample, we have provided a script to generate a self-signed device certification used for device authentication. This script is intended **for sample use only and not to be used in production code**.

Reviewer 2: nit: should this be a relative link?

Author: I can totally do that if you prefer.  I think I put in the full url because I could test the link before committing.

1. Enter the directory `/azure-sdk-for-c/sdk/samples/iot/provisioning/src/` and run the script using the following form:

    Linux:
    ```bash
    ./generate_certificate.sh
    ```
    Windows:
    ```cmd
    generate_certificates.cmd
    ```

> 2. Set the environment variable `AZ_IOT_DEVICE_X509_CERT_PEM_FILE` noted at the bottom of the output.

Reviewer: ```suggestion
2. Set the environment variable `AZ_IOT_DEVICE_X509_CERT_PEM_FILE` to the path of the generated .pem file noted at the bottom of the output.
```

3. In your Azure DPS, add a new individual device enrollment using the recently generated `device_ec_cert.pem` file.  See [here](https://docs.microsoft.com/en-us/azure/iot-dps/quick-create-simulated-device-x509#create-a-device-enrollment-entry-in-the-portal) for further instruction.  After creation, the Registration ID of your device should appear as `paho-sample-device1` in the Individual Enrollments tab.

### Paho IoT Provisioning (SAS)
This [sample](https://github.com/Azure/azure-sdk-for-c/blob/master/sdk/samples/iot/provisioning/src/paho_iot_provisioning_sas_example.c) uses SAS symmetric key authentication to connect to Azure IoT Hub Device Provisioning Service (DPS). When enrolling a device with the DPS, an option to auto-generate the key will be provided.  

* In your Azure DPS, add a new individual device enrollment using SAS. See [here](https://docs.microsoft.com/en-us/azure/iot-dps/quick-create-simulated-device-symm-key#create-a-device-enrollment-entry-in-the-portal) for further instruction. After creation, the Registration ID of your device will appear in the Individual Enrollments tab.

> 

Reviewer 2: nit: remove one of these newlines

Author: you don't want three?? 😆

## Build and Run the Sample

1. Set the remaining environment variables.  Setting a variable will take the following form:

	  Linux:
	  ```bash
	  export ENV_VARIABLE_NAME=VALUE
	  ```
	  Windows:
	  ```cmd
	  set "ENV_VARIABLE_NAME=VALUE"
	  ```
	#### Paho IoT Provisioning (Certificates)

>   * `VCPKG_DEFAULT_TRIPLET` and `VCPKG_ROOT`: These should already be set per these [directions](https://github.com/Azure/azure-sdk-for-c#development-environment).

Reviewer 2: relative link?

Reviewer 3: I forgot to bring this up in our meeting, but there are folks on our engsys team working towards fixing potential 404 or incorrect links in READMEs and docs across all the languages, and there recommendation was to use complete links rather than relative ones.

For example, see [this file change](https://github.com/Azure/azure-sdk-for-js/pull/9978/files#diff-f57514279a5dcfcfdf7059f4ec56e4a9R8) from https://github.com/Azure/azure-sdk-for-js/pull/9978 OR https://github.com/Azure/azure-sdk-for-python/pull/12543
cc @KarishmaGhiya

Author: i see.  ok thanks for the info ahson!

  * `AZ_IOT_DEVICE_X509_CERT_PEM_FILE`: This should already be set per directions above.
  * `AZ_IOT_ID_SCOPE`: Copy this value from the Overview tab in your Azure DPS.
  * `AZ_IOT_REGISTRATION_ID`: This should be `paho-sample-device1`.
	#### Paho IoT Provisioning (SAS)

>   * `VCPKG_DEFAULT_TRIPLET` and `VCPKG_ROOT`: These should already be set per these [directions](https://github.com/Azure/azure-sdk-for-c#development-environment).

Reviewer 2: relative

  * `AZ_IOT_PROVISIONING_SAS_KEY`: Select your enrolled device from the Individual Enrollments tab and copy its Primary Key.
  * `AZ_IOT_ID_SCOPE`: Copy this value from the Overview tab in your Azure DPS.
  * `AZ_IOT_REGISTRATION_ID_SAS`: Copy the Registration Id of your SAS device from the Individual Enrollments tab.

2. Compile the code:
  * Enter the directory `/azure-sdk-for-c/cmake`.  If it does not exist, please create it.
  * Build the directory structure:

    ```bash
    cmake -DTRANSPORT_PAHO=ON ..
    ```
  * Build the sample:

    Linux:
    ```bash

>     make

Reviewer 2: nit: I think this should stay `cmake --build .` It does the same thing as `make` and prevents confusion if someone decides to use Ninja etc.

Author: > Great work here. I'm usually pretty "nit"y with docs but this is awesome work. Thanks for all the docs!

Thanks Dane, I appreciate that!

    ```
    Windows:
    ```cmd
    cmake --build .
    ```
3. From within the cmake directory, run the sample:

    Linux:
    ```bash
    ./sdk/samples/iot/provisioning/<sample executable here>
    ```
    Windows:
    ```cmd
    az.sln
    ```
	Once the Windows solution opens in Visual Studio:
    * Navigate on the Solution Explorer panel to the sample solution you would like to run.
    * Make it the default startup project (right-click on sample project, then click on Set as StartUp Project).
    * Build and run the project (F5 on most installations).

## Troubleshooting

> The error policy in using the Embedded C SDK client library is documented [here](https://github.com/Azure/azure-sdk-for-c/blob/master/sdk/docs/iot/mqtt_state_machine.md#error-policy).

Reviewer 2: relative?

## Next Steps and Additional Documentation

Start using the IoT Provisioning Client in your solutions! 

> * A general overivew of the Embedded C SDK and additional background on runing samples can be found in the [Azure SDK for Embedded C README](https://github.com/Azure/azure-sdk-for-c#azure-sdk-for-embedded-c).

Reviewer 3: ```suggestion
* A general overview of the Embedded C SDK and additional background on running samples can be found in the [Azure SDK for Embedded C README](https://github.com/Azure/azure-sdk-for-c#azure-sdk-for-embedded-c).
```
